### PR TITLE
Add a header file including all the collections

### PR DESCRIPTION
**Pull request by jmcarcell**
BEGINRELEASENOTES
- Add a header file including all the collections in the datamodel. This header will be generated and installed into `<datamodel>/<datamodel>.h`.

ENDRELEASENOTES

The usage would be 

``` cpp
#include "edm4hep/edm4hep.h"
```

and all the collections will be included. 

I find from time to time I need to do something with all the collections and after having a look at other options (hardcoding the includes means they always have to be changed; having some CMake magic in EDM4hep to generate this file is an option but it also can be done at the podio level) I think this is the cleanest one and other people can maybe also use it. In practice, if an algorithm does something with the collections and a new one is added, even if the include doesn't change because it comes from this file other code that does something with collections will have to be changed.

### CODE REVIEW

**Review by tmadlener — COMMENTED** [review]
Looks reasonable to me. I am not sure if `AllCollections.h` is the best name, but I don't have a better one either. Maybe something that conveys better that this is the **complete datamodel**?

You also need to add the new template here to trigger re-generation in case the template changes: 
https://github.com/AIDASoft/podio/blob/master/python/templates/CMakeLists.txt

File: python/podio_gen/cpp_generator.py [hunk 1]
```diff
@@ -485,6 +486,22 @@ def _write_list(name, target_folder, files, comment):
...
         )
 
+    def _write_all_collections_header(self):
+        """Write a header file that includes all collection headers"""
+
+        collection_files = (
+            os.path.basename(f)
+            for f in self.generated_files
+            if f.endswith("Collection.h") and "Mutable" not in f
+        )
```
[the remark this answers is not in the document]
Comment (jmcarcell) [reply python/podio_gen/cpp_generator.py]: I changed it to use `self.datamodel.datatypes`

File: python/templates/AllCollections.h.jinja2 [hunk 2]
```diff
@@ -0,0 +1,5 @@
+// AUTOMATICALLY GENERATED FILE - DO NOT EDIT
+
```
[the remark this answers is not in the document]
Comment (jmcarcell) [reply python/templates/AllCollections.h.jinja2:2]: Added, although it should not matter much since the headers have their own

**Comment by jmcarcell** [comment]
> Looks reasonable to me. I am not sure if `AllCollections.h` is the best name, but I don't have a better one either. Maybe something that conveys better that this is the **complete datamodel**?
> 
> You also need to add the new template here to trigger re-generation in case the template changes: [`master`/python/templates/CMakeLists.txt](https://github.com/AIDASoft/podio/blob/master/python/templates/CMakeLists.txt?rgh-link-date=2024-05-24T07%3A01%3A47Z)

I changed it to, for example for EDM4hep, `edm4hep/Alledm4hepCollections.h`

**Comment by hegner** [comment]
Why don't we just call it `edm4hep/edm4hep.h` ?

**Comment by m-fila** [comment]
I think I prefer `edm4hep/AllCollections.h` over `edm4hep/Alledm4hepCollections.h` but just `edm4hep/Collections.h` reads even better for me
Name of this header technically becomes reserved for componenents/datatatypes/interfaces (for instance 'DatamodelDefinition' is already reserved as using it as a name for anything would collide with `DatamodelDefinition.h`). It's super formal and pedantic but a list of reserved keywords could be documented and potentially validated

> Maybe something that conveys better that this is the complete datamodel?

 `edm4hep/Datamodel.h`or  `edm4hep/edm4hep.h`? Wouldn't the components not used in any datatype be missing though?

**Comment by hegner** [comment]
> `edm4hep/Datamodel.h`or `edm4hep/edm4hep.h`? Wouldn't the components not used in any datatype be missing though?

the missing components won't be needed in any case. Maybe for downstream code that would then include explicitly anyhow.

**Comment by tmadlener** [comment]
As decided during the [EDM4hep meeting on Jun 4](https://indico.cern.ch/event/1423788/) the name of the header should be `datamodel/datamodel.h`.

**Comment by tmadlener** [comment]
I changed the name of the generated header and also changed the generator logic to ensure that it is installed, as I don't see a reason not to install it.

**Comment by jmcarcell** [comment]
Looks good

File: python/templates/CMakeLists.txt [hunk 3]
```diff
@@ -1,6 +1,7 @@
 set(PODIO_TEMPLATES
   ${CMAKE_CURRENT_LIST_DIR}/Collection.cc.jinja2
   ${CMAKE_CURRENT_LIST_DIR}/Collection.h.jinja2
+  ${CMAKE_CURRENT_LIST_DIR}/AllCollections.h.jinja2
```
Comment (hegner) [inline python/templates/CMakeLists.txt:4]: Only nitpicking comment would be to change the name of the jinja2 template as well

**Review by hegner — CHANGES_REQUESTED** [review]

File: python/templates/CMakeLists.txt (hunk 3 above)
Comment (tmadlener) [reply python/templates/CMakeLists.txt]: done.

**Review by tmadlener — APPROVED** [review]

**Pull request merged**